[mlir][ArmSME] Add rudimentary support for tile spills to the stack

MacDue opened the pull request:
This adds very basic (and inelegant) support for something like spilling and reloading tiles, if you use more SME tiles than physically exist.

This is purely implemented to prevent the compiler from aborting if a function uses too many tiles (i.e. due to bad unrolling), but is expected to perform very poorly.

Currently, this works in two stages:

During tile allocation, if we run out of tiles instead of giving up, we switch to allocating 'in-memory' tile IDs. These are tile IDs that start at 16 (which is higher than any real tile ID). A warning will also be emitted for each (root) tile op assigned an in-memory tile ID:

```
warning: failed to allocate SME virtual tile to operation, all tile operations will go through memory, expect degraded performance
```

Everything after this works like normal until `-convert-arm-sme-to-llvm`

Here the in-memory tile op:

```mlir
arm_sme.tile_op { tile_id = <IN MEMORY TILE> }
```

Is lowered to:

```mlir
// At function entry:
%alloca = memref.alloca ... : memref<?x?xty>

// Around the op:
// Swap the contents of %alloca and tile 0.
scf.for %slice_idx {
  %current_slice = "arm_sme.intr.read.horiz" ... <{tile_id = 0 : i32}>
  "arm_sme.intr.ld1h.horiz"(%alloca, %slice_idx)  <{tile_id = 0 : i32}>
  vector.store %current_slice, %alloca[%slice_idx, %c0]
}
// Execute op using tile 0.
arm_sme.tile_op { tile_id = 0 }
// Swap the contents of %alloca and tile 0.
// This restores tile 0 to its original state.
scf.for %slice_idx {
  %current_slice = "arm_sme.intr.read.horiz" ... <{tile_id = 0 : i32}>
  "arm_sme.intr.ld1h.horiz"(%alloca, %slice_idx)  <{tile_id = 0 : i32}>
  vector.store %current_slice, %alloca[%slice_idx, %c0]
}
```

This is inserted during the lowering to LLVM as spilling/reloading registers is a very low-level concept, that can't really be modeled correctly at a high level in MLIR.

Note: This is always doing the worst case full-tile swap. This could be optimized to only spill/load data the tile op will use, which could be just a slice. It's also not making any use of liveness, which could allow reusing tiles. But these is not seen as important as correct code should only use the available number of tiles.

MacDue committed:
[mlir][ArmSME] Add rudimentary support for tile spills to the stack

This adds very basic and inelegant support for something like spilling
and reloading tiles if you use more SME tiles than physically exist.

This is purely implemented to prevent the compiler from aborting if a
function uses too many tiles (i.e. due to bad unrolling), but is
expected to perform very poorly.

Currenly, this works in two stages:

During tile allocation, if we run out of tiles instead of giving up, we
switch to allocating 'in-memory' tile IDs. These are tile IDs that start
at 16 (which is higher than any real tile ID). A warning will also be
emitted for each (root) tile op assigned an in-memory tile ID:

```
warning: failed to allocate physical tile to operation, all tile operations will go through memory, expect performance degradation
```

Everything after this works like normal until `-convert-arm-sme-to-llvm`

Here the in-memory tile op:

```mlir
arm_sme.tile_op { tile_id = <IN MEMORY TILE> }
```

Is lowered to:

```mlir
// At function entry:
%alloca = memref.alloca ... : memref<?x?xty>

// Around the op:
// Swap the contents of %alloca and tile 0.
scf.for %slice_idx {
  %current_slice = "arm_sme.intr.read.horiz" ... <{tile_id = 0 : i32}>
  "arm_sme.intr.ld1h.horiz"(%alloca, %slice_idx)  <{tile_id = 0 : i32}>
  vector.store %current_slice, %alloca[%slice_idx, %c0]
}
// Execute op using tile 0.
arm_sme.tile_op { tile_id = 0 }
// Swap the contents of %alloca and tile 0.
// This restores tile 0 to its original state.
scf.for %slice_idx {
  %current_slice = "arm_sme.intr.read.horiz" ... <{tile_id = 0 : i32}>
  "arm_sme.intr.ld1h.horiz"(%alloca, %slice_idx)  <{tile_id = 0 : i32}>
  vector.store %current_slice, %alloca[%slice_idx, %c0]
}
```

This is inserted during the lowering to LLVM as spilling/reloading
registers is a very low-level concept, that can't really be modeled
correctly at a high level in MLIR.

Note: This is always doing the worst case full-tile swap. This could be
optimized to only spill/load data the tile op will use, which could be
just a slice. It's also not making any use of liveness, which could
allow reusing tiles. But these is not seen as important as correct code
should only use the available number of tiles.

c-rhodes reviewed (commented):
thanks for the patch Ben! I've not gone through it all yet, but left some comments for now. Will have another pass later, cheers

banach-space reviewed (commented):
Great work! Mostly makes sense, but I have a few small questions/suggestions.

From summary:

> Here the in-memory tile op:
> 
> arm_sme.tile_op { tile_id = <IN MEMORY TILE> }

Could you clarify what "IN MEMORY TILE" is in practice? IIUC, it's an integer >= 16?

> This could be optimized to only spill/load data the tile op will use, which could be just a slice.

Could you add a test that demonstrates excessive spilling?

> It's also not making any use of liveness, which could allow reusing tiles. But these is not seen as important as correct code should only use the available number of tiles.

Please document these as next steps for optimisation. In particular, this is important, but we may not have the bandwidth to prioritise this just yet. Functional correctness first, performance next ;-)

MacDue committed:
fixups

MacDue committed:
fixups

- Show alloca usage in tests
- Add test showing some very excessive spills
- Document a possible API to reduce spills

MacDue committed:
Remove newline

MacDue committed:
Fix build error

MacDue commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -548,6 +746,17 @@ struct ConvertArmSMEToLLVMPass
+template <typename... TileOp>
+static void addSpillAndFillsForTileOp(RewritePatternSet &patterns,
+                                      LLVMTypeConverter const &typeConverter) {
+  // Add spill/fill conversions with a very high benefit to ensure they are
+  // lowered first.
```
If the op is lowered to an intrinsic before the spills/fills are added, then there's no op to add spills/fills to.

MacDue commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -40,6 +132,209 @@ IntegerAttr getTileIdOrError(arm_sme::ArmSMETileOpInterface op) {
+  return alloca;
+}
+
+/// Finds or creates an alloca for a spill of a tile.
```
This is already checked via the first check in the `ConvertArmSMESpillsAndFillsToLLVM` rewrite (which looks at `tileOp.isInMemoryTile()`)

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:
```diff
@@ -246,16 +259,18 @@ struct AssignTileIDsPattern
-          return tileOp.emitOpError(
+          return dependantTileOp.emitOpError(
               "already assigned different SME virtual tile!");
-        rewriter.updateRootInPlace(tileOp,
```
Because `tileOp` is not the root op here. I renamed it to `dependantTileOp` to make that clearer.

MacDue commented on mlir/test/Conversion/ArmSMEToLLVM/tile-spills-and-fills.mlir:
```diff
@@ -0,0 +1,153 @@
+/// Then around the op:
+///
+/// // Swap contents of %tileAlloca and tile 0
+/// scf.for %sliceIdx ... {
+///   %currentSlice = arm_sme.intr.read.horiz {tile_id = 0}
+///   arm_sme.intr.ld1h.horiz %tileAlloca[%sliceIdx, %c0] {tile_id = 0}
+///   vector.store %currentSlice, %tileAlloca[%sliceIdx, %c0]
```
Yep, and because the op is `zero` the `arm_sme.intr.ld1h.horiz` is pointless too. It's still here because the spilling/filling logic does not know anything about the op (i.e. if it only writes to ZA, how much of the tile it uses, etc). I've updated the comment above the rewrite to show a possible `tileOp.getZaUsage()` API, which could be used to make the spills/fills less dumb.

MacDue commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/use-too-many-tiles.mlir:
```diff
@@ -0,0 +1,77 @@
+  return
+}
+
+func.func @get_svl() -> index  attributes { enable_arm_streaming_ignore, arm_locally_streaming }{
```
I know, I'm just being explicit as I've manually set the streaming mode to ensure this returns SVL.

c-rhodes replied:
ah I see, perhaps there's value in us adding an op for this at some point, there's synthetic intrinsics in the backend for this the ACLE uses: https://arm-software.github.io/acle/main/acle.html#rdsvl

- int_aarch64_sme_cntsb
- int_aarch64_sme_cntsh
- int_aarch64_sme_cntsw
- int_aarch64_sme_cntsd

MacDue replied:
Those would probably be nicer to use than the current `vscale * base_size` we're currently doing :) 
Especially if we had a op like:
```mlir
%svl = arm_sme.streaming_vl <bytes|half-words|words|double-words>
```

MacDue commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -40,6 +132,204 @@ IntegerAttr getTileIdOrError(arm_sme::ArmSMETileOpInterface op) {
+///
+///     // Around op:
+///     scf.for %slice_idx {
+///       %current_slice = "arm_sme.intr.read.horiz" ... <{tile_id = 0 : i32}>
+///       "arm_sme.intr.ld1h.horiz"(%spill, %slice_idx)  <{tile_id = 0 : i32}>
+///       vector.store %current_slice, %spill[%slice_idx, %c0]
```
These can't be reordered as the `vector.store` overwrites the slice loaded by the `arm_sme.intr.ld1h.horiz`. Remember this is doing an in-place swap.

c-rhodes replied:
ah of course 🤦

MacDue committed:
fixups

c-rhodes commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -129,6 +132,221 @@ IntegerAttr getTileIdOrError(arm_sme::ArmSMETileOpInterface op) {
+    // touches (i.e. a single tile slice).
+    {
+      rewriter.setInsertionPoint(op);
+      // Swap the in-memory tile's contents into ZA before the op.
```
nit: could be a little more descriptive
```suggestion
      // Swap the contents of ZA and the in-memory tile before the op.
```

c-rhodes commented on mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:
```diff
@@ -209,28 +210,40 @@ struct AssignTileIDsPattern
+    if (!tileIsInMemory)
+      setDiscardableIntAttr(kTilesInUseAttr, tilesInUse);
+    else {
+      // If we could not find a real tile, set use a virtual tile ID (ID >= 16).
```
```suggestion
      // If we could not find a real tile, use a virtual tile ID (ID >= 16).
```

c-rhodes commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -129,6 +132,221 @@ IntegerAttr getTileIdOrError(arm_sme::ArmSMETileOpInterface op) {
+///       vector.store %slice_to_save, %spill[%slice_idx, %c0]
+///     }
+///
+/// Note that these spills/fills are not inserted earlier as concept of a
+/// register, and the need to swap the contents, can't really be represented
+/// correctly at a high level in MLIR.
```
could this be done as a separate transform pass (like TileAllocation.cpp) that's run before `-arm-sme-to-llvm`? It feels a bit odd to do this during conversion, especially since this has to be registered for specific ops (unlike greedy rewriter):
```
  // Register ops that need spills/fills.
  addSpillAndFillsForTileOp<
      arm_sme::LoadTileSliceOp, arm_sme::MoveTileSliceToVectorOp,
      arm_sme::MoveVectorToTileSliceOp, arm_sme::StoreTileSliceOp,
      arm_sme::OuterProductOp, arm_sme::ZeroOp>(patterns, converter);
```

and we must ensure new ops that also need spills/fills are added here, which could easily be missed?

MacDue replied:
It's specifically done here so it's possible to use the intrinsics (which needs help from the LLVM converter to get a memref ptr), and because this concept really does not make sense at a higher level than intrinsics. I don't want pass that puts the IR in a weird state that mixes intrinsics and high-level ops.

There's really no where I think this concept makes much sense in MLIR, because it's all based around SSA values.

MacDue replied:
I've now made adding the spills and fills conversion opt-out (rather than op-in) see: https://github.com/llvm/llvm-project/pull/76086/commits/d95777bd9adb833f546d545b2282038945788cb0.

MacDue committed:
Make adding spills/fills conversions implict

This switches the ArmSME -> LLVM conversion patterns to use a new
`ConvertArmSMEOpToLLVMPattern` base class. Using this implicitly adds
the spills/fills conversion pattern, unless the `requiresSpillsAndFills`
template parameteris set to `RequiresSpillsAndFills::No`.

MacDue committed:
fixups

MacDue committed:
Remove space

MacDue committed:
Move helper

MacDue committed:
Remove redundant template parameters

banach-space commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -129,8 +132,267 @@ IntegerAttr getTileIdOrError(arm_sme::ArmSMETileOpInterface op) {
+
+  /// Emits a full in-place swap of the contents of a tile in ZA and a
+  /// tile-sized memref (`tileAlloca`).
+  void emitFullTileSwap(RewriterBase &rewriter, Location loc, Value tileAlloca,
```
I think that removing the final argument (which is hard-coded in practice) and renaming this as e.g. `swapInMemoryTileWithSMETileZero` could help to convey the fact that right now everything is hard-coded to use Tile 0.

banach-space commented on mlir/test/Conversion/ArmSMEToLLVM/tile-spills-and-fills.mlir:
```diff
@@ -0,0 +1,153 @@
+/// Checks tile spill/reloads are inserted around in-memory tiles (i.e. tiles
+/// that were not assigned a physical SME tile).
+///
+/// These spills are currently very naive and paranoid and will spill/reload
```
[nit] Remove paranoid - doesn't add any detail that's not already present in this comment

banach-space commented on mlir/test/Conversion/ArmSMEToLLVM/tile-spills-and-fills.mlir:
```diff
@@ -0,0 +1,153 @@
+/// }
+///
+
+// -----
```
DELETEME

banach-space commented on mlir/test/Conversion/ArmSMEToLLVM/tile-spills-and-fills.mlir:
```diff
@@ -0,0 +1,153 @@
+//  AFTER-LLVM-LOWERING-DAG: %[[TILE_ALLOCA:.*]] = memref.alloca(%[[SVL_H]], %[[SVL_H]])
+// AFTER-LLVM-LOWERING-SAME:   {arm_sme.in_memory_tile_id = 16 : i32} : memref<?x?xi16>
+//
+//  AFTER-LLVM-LOWERING-NOT: scf.for
+//      AFTER-LLVM-LOWERING: arm_sme.intr.zero
+//
+//  AFTER-LLVM-LOWERING-NOT: scf.for
+//      AFTER-LLVM-LOWERING: arm_sme.intr.zero
```
In addition to this, why not check for whatever `arm_sme.zero` is lowered to?

banach-space reviewed (approved):
This is quite involved, but very well crafted and actually relatively easy to follow. I've left a few minor comments, but otherwise LGTM!

Btw, you've written a lot of useful documentation while working on this - could you summarise this in https://mlir.llvm.org/docs/Dialects/ArmSME/? That shouldn't be too much work given that you've already done most of it :)

MacDue committed:
Fixup

MacDue committed:
Fixups

c-rhodes reviewed (approved):
LGTM cheers Ben!

I know you mentioned it's not exactly optimal or elegant, and I agree, but we have to play the hand we're dealt and in this case it's a set of intrinsics that force us to do things we probably shouldn't be doing in MLIR. Hopefully a bit further down the road these intrinsics will either evolve or a new set will come along where the tile is a proper type with spilling/filling support in the backend and dataflow is correctly modeled.

In the meantime, the functionality you've implemented here allows code that uses more tiles than exist at the architecture level to correctly run and there's value in that. It's much better than a crash anyway! 😄

ronlieb commented:
hi, we have SLES 15 SP4 systems with gcc 7.5  (llvm minimum supported version of gnu is 7.4) and see build time failures with this patch applied.

RHEL8 and ubuntu20 compile code ok,however much more recent gnu toolchain.

llvm-project/mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp: In lambda function:
llvm-project/mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:377:67: error: parameter packs not expanded with â€˜...â€™:
         if constexpr (Pattern::requiresSpillsAndFillsConversion() &&
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
                       std::is_base_of_v<arm_sme::ArmSMETileOpInterface::Trait<
                       ~~~

MacDue commented:
@ronlieb Hi, could you test if https://github.com/llvm/llvm-project/pull/78046 fixes the issue? I tried something similar compiler explorer and it seemed to work for GCC 7.4, but I don't have such a old compiler setup to build LLVM locally.

ronlieb commented:
> @ronlieb Hi, could you test if #78046 fixes the issue? I tried something similar compiler explorer and it seemed to work for GCC 7.4, but I don't have such an old compiler setup to build LLVM locally.

thx for fix, i will give it a go, it will need to grind through our CI, so likely will take about 7 hours

ronlieb commented:
> > @ronlieb Hi, could you test if #78046 fixes the issue? I tried something similar compiler explorer and it seemed to work for GCC 7.4, but I don't have such an old compiler setup to build LLVM locally.
> 
> thx for fix, i will give it a go, it will need to grind through our CI, so likely will take about 7 hours

the patch worked, thanks!  SLES build succeeded

MacDue commented:
> > > @ronlieb Hi, could you test if #78046 fixes the issue? I tried something similar compiler explorer and it seemed to work for GCC 7.4, but I don't have such an old compiler setup to build LLVM locally.
> > 
> > thx for fix, i will give it a go, it will need to grind through our CI, so likely will take about 7 hours
> 
> the patch worked, thanks!  SLES build succeeded

Landed the fix now, thanks for testing! 👍
